akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -137,9 +137,14 @@ import scala.concurrent.{ Future, Promise }
 
         def onPull(): Unit = pull(in)
 
-        override def onDownstreamFinish(): Unit = {
-          finishPromise.success(Done)
-          completeStage()
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          cause match {
+            case _: SubscriptionWithCancelException.NonFailureCancellation =>
+              finishPromise.success(Done)
+            case ex =>
+              finishPromise.failure(ex)
```

Comment:
That's a change that we need to document, I guess?

Reply:
Or it should always complete regardless of why downstream cancels?

Reply:
Documented